hoomd/mpcd/RandomSubsetPicker.cu

```diff
@@ -0,0 +1,184 @@
+ *
+ * The CUB library is used to compact the particle indexes of the selected particles
+ * into \a d_flags_idx based on the flags set in \a d_flags. The number of flagged
+ * particles is also determined.
+ *
+ * \note This function must be called twice. On the first call, the temporary storage
+ *       required is sized and stored in \a tmp_storage_bytes. Device memory must be
+ *       allocated to \a d_tmp_storage, and then the function can be called again
+ *       to apply the transformation.
+ *
+ * \note Per CUB user group, DeviceSelect is in-place safe, and so input and output
+ *       do not require a double buffer.
+ *
+ * See kernel::create_flags_idx for details of how particles indices are stored
+ */
+template<typename T>
+cudaError_t compact_flags_idx(unsigned int *d_flags_idx,
+                              unsigned int *d_num_flags,
+                              void *d_tmp_storage,
+                              size_t &tmp_storage_bytes,
+                              T *d_flags,
+                              unsigned int N)
+    {
+    if (N == 0) return cudaSuccess;
+
+    HOOMD_CUB::DeviceSelect::Flagged(d_tmp_storage, tmp_storage_bytes, d_flags_idx, d_flags, d_flags_idx, d_num_flags, N);
+
+    return cudaSuccess;
+    }
+
+/*!
+ * \param d_picks_idx Indexes of picked particles in \a d_flags_idx
+ * \param d_picks Indexes of picked particles in \a d_flags
+ * \param d_flags_idx Compacted array of particle indexes which has *1* in flags
+ * \param N_pick Number of picks made
+ * \param block_size Number of threads per block
+ *
+ * \sa kernel::store_picks_idx
+ */
+cudaError_t store_picks_idx(unsigned int *d_picks_idx,
+                            unsigned int *d_picks,
+                            unsigned int *d_flags_idx,
+                            unsigned int N_pick,
+                            unsigned int block_size)
+    {
+    if (N_pick == 0) return cudaSuccess;
+
+    static unsigned int max_block_size = UINT_MAX;
+    if (max_block_size == UINT_MAX)
+        {
+        cudaFuncAttributes attr;
+        cudaFuncGetAttributes(&attr, (const void*)kernel::store_picks_idx);
+        max_block_size = attr.maxThreadsPerBlock;
+        }
+
+    const int run_block_size = min(block_size, max_block_size);
+    kernel::store_picks_idx<<<N_pick/run_block_size+1, run_block_size>>>(d_picks_idx,
+                                                                         d_picks,
+                                                                         d_flags_idx,
+                                                                         N_pick);
+
+    return cudaSuccess;
+    }
+
+//! Template instantiation of unsigned int
+template cudaError_t compact_flags_idx<unsigned int>
+    (unsigned int *d_flags_idx,
+     unsigned int *d_num_flags,
+     void *d_tmp_storage,
+     size_t &tmp_storage_bytes,
+     unsigned int *d_flags,
+     unsigned int N);
+
+//! Template instantiation of unsigned char
```

Comment:
Is there a spot where we use this  for `unsigned char`? We should only explicitly instantiate the templates we actually use.

Reply:
No we didn't, removed the template instantiate for unsigned char